### PR TITLE
runtime updates

### PR DESCRIPTION
**Pull request by amarzavery**
- added support for polymorphicDiscriminator and tests for it.
-  logger for interactiveLogin
-  removed underscore dependency and added a retry filter to handle system errors

### CODE REVIEW

File: runtime/ms-rest/lib/filters/systemErrorRetryPolicyFilter.js [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+ */
+function SystemErrorRetryPolicyFilter(retryCount, retryInterval, minRetryInterval, maxRetryInterval) {
+  function newFilter(options, next, callback) {
+    var retryData = null;
+
+    function retryCallback(err, response, body) {
+      retryData = newFilter.updateRetryData(retryData, err);
+      if (!utils.objectIsNull(err) && newFilter.shouldRetry(retryData, err) && 
+          (err.code === 'ETIMEDOUT' || err.code === 'ESOCKETTIMEDOUT' || err.code === 'ECONNREFUSED' || 
+           err.code === 'ECONNRESET' || err.code === 'ENOTFOUND')) {
```
Comment (amarzavery) [inline runtime/ms-rest/lib/filters/systemErrorRetryPolicyFilter.js]: @huangpf - This is the new filter that will retry for some of the selected system Errors. This should solve the issue reported here in xplat cli https://github.com/Azure/azure-xplat-cli/issues/3073

Comment (amarzavery) [inline runtime/ms-rest/lib/filters/systemErrorRetryPolicyFilter.js]: Please review the PR and let me know if there are any concerns.

File: runtime/ms-rest/lib/serviceClient.js [hunk 2]
```diff
@@ -52,6 +53,7 @@ function ServiceClient(credentials, options) {
   options.filters.push(RedirectFilter.create());
   if (!options.noRetryPolicy) {
     options.filters.push(new ExponentialRetryPolicyFilter());
+    options.filters.push(new SystemErrorRetryPolicyFilter());
```
Comment (amarzavery) [inline runtime/ms-rest/lib/serviceClient.js]: @huangpf - the retry filter is added by default. So no one needs to add extra code unless they are not satisfied by the default scenario.

File: runtime/ms-rest/test/retryPolicyTests.js [hunk 3]
```diff
@@ -131,3 +132,165 @@ describe('exponentialretrypolicyfilter-unittests', function () {
     });
   });
 });
+
+describe('systemErrorRetrypolicyfilter-unittests', function () {
+  it('RetrySucceedsOn_ENOTFOUND_ErrorCode', function (done) {
```
Comment (amarzavery) [inline runtime/ms-rest/test/retryPolicyTests.js]: @huangpf - these are the unit tests that ensure that the systemErrorRetryFilter works as expected for sleect system errors.

**Comment by huangpf** [comment]
LGTM. Thanks~

**Pull request merged**